servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpRequestAutoDrainingServiceFilter.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright © 2025 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.SingleSource;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.concurrent.api.TerminalSignalConsumer;
+import io.servicetalk.concurrent.internal.CancelImmediatelySubscriber;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+import io.servicetalk.http.utils.BeforeFinallyHttpOperator;
+
+import javax.annotation.Nullable;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.servicetalk.concurrent.api.Single.defer;
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.http.api.HttpExecutionStrategies.offloadNone;
+
+public final class HttpRequestAutoDrainingServiceFilter implements StreamingHttpServiceFilterFactory {
+
+    static final StreamingHttpServiceFilterFactory INSTANCE = new HttpRequestAutoDrainingServiceFilter();
+
+    private HttpRequestAutoDrainingServiceFilter() {
+        // singleton
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new DrainingStreamingHttpServiceFilter(service);
+    }
+
+    private static final class DrainingStreamingHttpServiceFilter extends StreamingHttpServiceFilter {
+
+        DrainingStreamingHttpServiceFilter(StreamingHttpService service) {
+            super(service);
+        }
+
+        @Override
+        public Single<StreamingHttpResponse> handle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            return defer(() -> doHandle(ctx, request, responseFactory).shareContextOnSubscribe());
```

Comment:
Continuation of previous discussion: https://github.com/apple/servicetalk/pull/3228#discussion_r2035937063

The main point there was that we only need to drain once to free up connection state in `NettyChannelPublisher` and complete `requestCompletion` in `NettyHttpServer`, doesn't matter if users add retries or not or if they do something else. I don't see what what else users can do to break something here.

Another benefit of removing `defer` will be an ability to use static atomic updater and avoid allocation of `AtomicBoolean` on every request.

Reply:
I not sure why we need to remove the defer to add our CAS to the terminal signal consumer object: we can do that either way.
I don't have really strong opinions about the `defer(..)`, we can leave it non-defered for now and if we ever run into a problem in the future we can add it back in.

Reply:
Reasoning is that this is a one-time operation per request rather than something we need to redo on every potential re-subscribe that can be added only by users (we always subscribe once)